src/libstore/aws-creds.cc

```diff
@@ -0,0 +1,261 @@
+#include "nix/store/aws-creds.hh"
+
+#if NIX_WITH_S3_SUPPORT
+
+#  include <aws/crt/Types.h>
+#  include "nix/store/s3-url.hh"
+#  include "nix/util/finally.hh"
+#  include "nix/util/logging.hh"
+#  include "nix/util/sync.hh"
+#  include "nix/util/url.hh"
+#  include "nix/util/util.hh"
+
+#  include <aws/crt/Api.h>
+#  include <aws/crt/auth/Credentials.h>
+#  include <aws/crt/io/Bootstrap.h>
+
+#  include <boost/unordered/concurrent_flat_map.hpp>
+
+#  include <chrono>
+#  include <condition_variable>
+#  include <mutex>
+#  include <unistd.h>
+
+namespace nix {
+
+namespace {
+
+// AWS CRT initialization
+static bool initAwsCrt()
```

Comment:
Now that I see the callers, we can actually go back to void, but different than before.

They all throw exceptions with a slightly different message.

Search around for the "catch, add trace, rethrow" pattern. This method can throw a nix error, and then the callers can do that to add contextual information.

Reply:
Gave it a shot in 09142ab, let me know if it matches what you wanted :)